src/main/java/vendored/com/google/common/util/concurrent/CycleDetectingLockFactory.java

```diff
@@ -0,0 +1,947 @@
+            implements CycleDetectingLock {
+
+        // These ReadLock/WriteLock implementations shadow those in the
+        // ReentrantReadWriteLock superclass. They are simply wrappers around the
+        // internal Sync object, so this is safe since the shadowed locks are never
+        // exposed or used.
+        private final CycleDetectingReentrantReadLock readLock;
+        private final CycleDetectingReentrantWriteLock writeLock;
+
+        private final LockGraphNode lockGraphNode;
+
+        private CycleDetectingReentrantReadWriteLock(LockGraphNode lockGraphNode, boolean fair) {
+            super(fair);
+            this.readLock = new CycleDetectingReentrantReadLock(this);
+            this.writeLock = new CycleDetectingReentrantWriteLock(this);
+            this.lockGraphNode = requireNonNull(lockGraphNode);
+        }
+
+        ///// Overridden ReentrantReadWriteLock methods. /////
+
+        @Override
+        public ReadLock readLock() {
+            return readLock;
+        }
+
+        @Override
+        public WriteLock writeLock() {
+            return writeLock;
+        }
+
+        ///// CycleDetectingLock methods. /////
+
+        @Override
+        public LockGraphNode getLockGraphNode() {
+            return lockGraphNode;
+        }
+
+        @Override
+        public boolean isAcquiredByCurrentThread() {
+            return isWriteLockedByCurrentThread() || getReadHoldCount() > 0;
+        }
+    }
+
+    private class CycleDetectingReentrantReadLock extends ReentrantReadWriteLock.ReadLock {
+
+        final CycleDetectingReentrantReadWriteLock readWriteLock;
+
+        CycleDetectingReentrantReadLock(CycleDetectingReentrantReadWriteLock readWriteLock) {
+            super(readWriteLock);
+            this.readWriteLock = readWriteLock;
+        }
+
+        @Override
+        public void lock() {
+            aboutToAcquire(readWriteLock);
+            try {
+                super.lock();
+            } finally {
+                lockStateChanged(readWriteLock);
+            }
+        }
+
+        @Override
+        public void lockInterruptibly() throws InterruptedException {
+            aboutToAcquire(readWriteLock);
+            try {
+                super.lockInterruptibly();
+            } finally {
+                lockStateChanged(readWriteLock);
+            }
+        }
+
+        @Override
+        public boolean tryLock() {
+            aboutToAcquire(readWriteLock);
+            try {
+                return super.tryLock();
+            } finally {
+                lockStateChanged(readWriteLock);
+            }
+        }
+
+        @Override
+        public boolean tryLock(long timeout, TimeUnit unit) throws InterruptedException {
+            aboutToAcquire(readWriteLock);
+            try {
+                return super.tryLock(timeout, unit);
+            } finally {
+                lockStateChanged(readWriteLock);
+            }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 871:876, 930:935.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.